Add recent downloads page for STS requests

davelopez opened the pull request:
This PR introduces a new **"Recent Downloads"** page that lists all Short Term Storage (STS) download requests made by the user. It leverages the existing [persistentProgressMonitor](https://github.com/galaxyproject/galaxy/blob/dev/client/src/composables/persistentProgressMonitor.ts) composable to keep track of the status of each request.

## Key Features:
- View recent STS requests: quickly see all your recent exports/downloads.

- Re-download available items: easily access files still available in the STS cache.

- Navigate to source objects: jump directly to the details page of the originating item (e.g., workflow invocation, history, etc.).

This improves the discoverability and accessibility of recent downloads without needing to track links manually.


![image](https://github.com/user-attachments/assets/a741a24f-70d4-476f-8438-84f7b32a4a1c)



Closes #20132

## TODO
- [x] Track history exports to STS
- [x] Add some tests
- [ ] Try to poll only for one item per poll cycle

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Export a workflow invocation or a history by selecting the "Temporary Download Link" option
  - Go to the new activity "Recent Downloads"

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Restrict ProcessedObject type to known values

davelopez committed:
Add workflow name to invocation export request data

To make more useful the information displayed in other components consuming the request data.

davelopez committed:
Add download tracker composable

Other components can now track when a new download/export to the STS is requested.

davelopez committed:
Integrate download tracking in PersistentTaskProgressMonitorAlert

davelopez committed:
Add Recent Downloads page and integrate with routing

davelopez committed:
Add Recent Downloads activity

davelopez committed:
Add "Copy Link" action to DownloadItemCard

davelopez committed:
Add download URL retrieval to prepareObjectDownload function

davelopez committed:
Integrate recent download tracking for histories

davelopez committed:
Sort download monitoring data by most recent first

davelopez committed:
Fix date format for request date in DownloadItemCard

Use ISO string as expected by GCard

davelopez committed:
Add detailed return type for usePersistentProgressTaskMonitor composable

davelopez committed:
Add unit tests for DownloadItemCard component

And fix some conditions detected while testing

davelopez committed:
Add useRoundRobinPoller composable for generic round-robin polling of items

davelopez committed:
Add isFinal property to MonitoringData

And update task monitoring logic

davelopez committed:
Add fetchTaskStatus method to TaskMonitor interface

and update related tests

davelopez committed:
Use round robin polling in recent downloads

davelopez committed:
Add stopWaitingForTask method to TaskMonitor

davelopez committed:
Refactor download tracking to use new storeProgressData function

To simplify request handling

davelopez committed:
Add stopMonitoring function to useShortTermStorage

davelopez committed:
Add cleanup logic on HistoryExport unmount

To stop task monitoring

davelopez committed:
Add stopCheckingStatus on DownloadItemCard

To ensure all polling stops.

davelopez committed:
Rename useRoundRobinPoller composable to useRoundRobinSelector

As the composable itself doesn't poll, justs selects the next item to "process".

davelopez committed:
Add cleanup logic on InvocationExportWizard

To stop any potential task monitoring

davelopez committed:
Add keepPolling option to fetchTaskStatus in useGenericMonitor

davelopez committed:
Fix start function to only set interval if items are available

davelopez committed:
Refactor useRoundRobinSelector

To improve state management and simplify logic

davelopez committed:
Make downloadMonitoringData readonly to prevent external mutations

davelopez committed:
Refactor next function to handle single item case

davelopez committed:
Simulate isRunning based on the current status

To avoid waiting until the corresponding round robin cycle to show the status when there are multiple active downloads.

davelopez committed:
Refactor useRoundRobinSelector

To accept a broader range of item sources

davelopez committed:
Rename useRoundRobinSelector.ts file to roundRobinSelector.ts

For consistency with other composables.

davelopez committed:
Add unit tests for useRoundRobinSelector composable

itisAliRH commented on client/src/components/Downloads/RecentDownloads.vue:
```diff
@@ -0,0 +1,68 @@
+</script>
+<template>
+    <div class="recent-downloads">
+        <Heading h1 separator inline size="lg">Recent Exports & Downloads</Heading>
```
Suggestion: Replace the current `<Heading>` with the existing `BreadcrumbHeading` component for consistency.

davelopez replied:
Something like https://github.com/galaxyproject/galaxy/pull/20585/commits/d5b5ac9cbf6db78641a4683f9d772de34965a689?
I thought we would use the BreadcrumbHeading only when we require drilling down :thinking:
But I guess it is more consistent if we want to do it in the future :+1: 
By the way, will it work for navigating to the items and then back to the downloads? Not sure how to set that up.

itisAliRH replied:
Right, thank you @davelopez!

Currently, if the item has `to` property, it renders as an active path. #20618 is fixing this.
<img width="503" alt="image" src="https://github.com/user-attachments/assets/1147afd7-f7a5-40c5-811a-0f3f5803031d" />

BreadcrumbHeading is designed to maintain a consistent heading across every page, and every page needs to have its items defined to navigate back to the top levels.

davelopez committed:
Replace Heading component with BreadcrumbHeading

davelopez committed:
Update Activity button title

From "Recent Downloads" to "Recent Exports" to avoid confusion with other types of "downloads"

itisAliRH reviewed (approved):
Looks good to me, thanks @davelopez!
